File: samples/AdventureWorksLT.WpfApp/SalesOrderWindow.Presenter.cs
using DevZest.Data.Presenters;
using DevZest.Data.Views;
using DevZest.Data;
using System.Windows;
using System;
using System.Threading.Tasks;
using System.Threading;

namespace DevZest.Samples.AdventureWorksLT
{
    partial class SalesOrderWindow
    {
        public static class Styles
        {
            public static readonly StyleId DataSheet = new StyleId(typeof(SalesOrderWindow));
        }

        private class Presenter : DataPresenter<SalesOrderInfo>, ForeignKeyBox.ILookupService
        {
            public Presenter(Window ownerWindow, AddressLookupPopup addressLookupPopup)
            {
                _ownerWindow = ownerWindow;
                _addressLookupPopup = addressLookupPopup;
            }

            private Window _ownerWindow;
            private AddressLookupPopup _addressLookupPopup;
            private RowBinding<ForeignKeyBox> _shipToAddressBinding;
            private RowBinding<ForeignKeyBox> _billToAddressBinding;
            private RowBinding<DataView> _subFormBinding;

            protected override void BuildTemplate(TemplateBuilder builder)
            {
                _subFormBinding = _.SalesOrderDetails.BindToDataView(() => new DetailPresenter(_ownerWindow)).WithStyle(Styles.DataSheet);
                var lineCountValidation = _.LineCount.BindToValidationPlaceholder(_subFormBinding);
                builder.GridRows("Auto", "*", "Auto")
                    .GridColumns("580")
                    .AddBinding(0, 0, _.BindToSalesOrderHeaderBox(IsNew, out _shipToAddressBinding, out _billToAddressBinding))
                    .AddBinding(0, 1, lineCountValidation)
                    .AddBinding(0, 1, _subFormBinding)
                    .AddBinding(0, 2, _.BindToSalesOrderFooterBox());
            }

            public DetailPresenter CurrentRowDetailPresenter
            {
                get { return (DetailPresenter)_subFormBinding[CurrentRow].DataPresenter; }
            }

            bool ForeignKeyBox.ILookupService.CanLookup(CandidateKey foreignKey)
            {
                if (foreignKey == _.FK_Customer)
                    return true;
                else if (foreignKey == _.FK_BillToAddress)
                    return true;
                else if (foreignKey == _.FK_ShipToAddress)
                    return true;
                else
                    return false;
            }

            void ForeignKeyBox.ILookupService.BeginLookup(ForeignKeyBox foreignKeyBox)
            {
                if (foreignKeyBox.ForeignKey == _.FK_Customer)
                {
                    var dialogWindow = new CustomerLookupWindow();
                    dialogWindow.Show(_ownerWindow, foreignKeyBox, CurrentRow.GetValue(_.CustomerID), _shipToAddressBinding[CurrentRow], _billToAddressBinding[CurrentRow]);
                }
                else if (foreignKeyBox.ForeignKey == _.FK_ShipToAddress || foreignKeyBox.ForeignKey == _.FK_BillToAddress)
                    BeginLookupAddress(foreignKeyBox);
                else
                    throw new NotSupportedException();
            }

            private void BeginLookupAddress(ForeignKeyBox foreignKeyBox)
            {
                var foreignKey = (Address.PK)foreignKeyBox.ForeignKey;
                if (_addressLookupPopup.FK == foreignKey)
                    _addressLookupPopup.IsOpen = false;
                else
                {
                    var customerID = CurrentRow.GetValue(_.CustomerID);
                    if (customerID.HasValue)
                    {
                        var addressID = foreignKeyBox.ForeignKey == _.FK_ShipToAddress ? _.ShipToAddressID : _.BillToAddressID;
                        _addressLookupPopup.Show(foreignKeyBox, CurrentRow.GetValue(addressID), customerID.Value);
                    }
                }
            }

            public int SalesOrderId
            {
                get { return _.SalesOrderID[0].Value; }
            }

            public bool IsNew
            {
                get { return SalesOrderId < 1; }
            }

            public override bool SubmitInput(bool focusToErrorInput = true)
            {
                if (!base.SubmitInput(focusToErrorInput))
                    return false;

                var detailsPresenter = _subFormBinding[CurrentRow].DataPresenter;
                return detailsPresenter.SubmitInput(focusToErrorInput);
            }

            public async Task<int?> SaveToDb(CancellationToken ct)
            {
                if (IsNew)
                    return await App.ExecuteAsync(db => db.CreateSalesOrderAsync(DataSet, ct));
                else
                {
                    await App.ExecuteAsync(db => db.UpdateSalesOrderAsync(DataSet, ct));
                    return null;
                }
            }
        }
    }
}

File: samples.vb/AdventureWorksLT.WpfApp/SalesOrderWindow.Presenter.vb
Imports System.Threading
Imports DevZest.Data
Imports DevZest.Data.Presenters
Imports DevZest.Data.Views

Partial Class SalesOrderWindow
    Public NotInheritable Class Styles
        Public Shared ReadOnly DataSheet As New StyleId(GetType(SalesOrderWindow))
    End Class

    Private Class Presenter
        Inherits DataPresenter(Of SalesOrderInfo)
        Implements ForeignKeyBox.ILookupService
        Public Sub New(ownerWindow As Window, addressLookupPopup As AddressLookupPopup)
            _ownerWindow = ownerWindow
            _addressLookupPopup = addressLookupPopup
        End Sub

        Private ReadOnly _ownerWindow As Window
        Private _addressLookupPopup As AddressLookupPopup
        Private _shipToAddressBinding As RowBinding(Of ForeignKeyBox)
        Private _billToAddressBinding As RowBinding(Of ForeignKeyBox)
        Private _subFormBinding As RowBinding(Of DataView)

        Protected Overrides Sub BuildTemplate(builder As TemplateBuilder)
            _subFormBinding = Entity.SalesOrderDetails.BindToDataView(Function() New DetailPresenter(_ownerWindow)).WithStyle(Styles.DataSheet)
            Dim lineCountValidation = Entity.LineCount.BindToValidationPlaceholder(_subFormBinding)
            Dim e = Entity
            builder.GridRows("Auto", "*", "Auto") _
                .GridColumns("580") _
                .AddBinding(0, 0, e.BindToSalesOrderHeaderBox(IsNew, _shipToAddressBinding, _billToAddressBinding)) _
                .AddBinding(0, 1, lineCountValidation) _
                .AddBinding(0, 1, _subFormBinding) _
                .AddBinding(0, 2, e.BindToSalesOrderFooterBox())
        End Sub

        Public ReadOnly Property CurrentRowDetailPresenter() As DetailPresenter
            Get
                Return CType(_subFormBinding(CurrentRow).DataPresenter, DetailPresenter)
            End Get
        End Property

        Function CanLookup(foreignKey As CandidateKey) As Boolean Implements ForeignKeyBox.ILookupService.CanLookup
            If foreignKey Is Entity.FK_Customer Then
                Return True
            ElseIf foreignKey Is Entity.FK_BillToAddress Then
                Return True
            ElseIf foreignKey Is Entity.FK_ShipToAddress Then
                Return True
            Else
                Return False
            End If
        End Function

        Sub BeginLookup(foreignKeyBox As ForeignKeyBox) Implements ForeignKeyBox.ILookupService.BeginLookup
            If foreignKeyBox.ForeignKey Is Entity.FK_Customer Then
                Dim dialogWindow As New CustomerLookupWindow()
                dialogWindow.Show(_ownerWindow, foreignKeyBox, CurrentRow.GetValue(Entity.CustomerID), _shipToAddressBinding(CurrentRow), _billToAddressBinding(CurrentRow))
            ElseIf foreignKeyBox.ForeignKey Is Entity.FK_ShipToAddress OrElse foreignKeyBox.ForeignKey Is Entity.FK_BillToAddress Then
                BeginLookupAddress(foreignKeyBox)
            Else
                Throw New NotSupportedException()
            End If
        End Sub

        Private Sub BeginLookupAddress(foreignKeyBox As ForeignKeyBox)
            Dim foreignKey As Address.PK = CType(foreignKeyBox.ForeignKey, Address.PK)
            If _addressLookupPopup.FK Is foreignKey Then
                _addressLookupPopup.IsOpen = False
            Else
                Dim customerID = CurrentRow.GetValue(Entity.CustomerID)
                If customerID.HasValue Then
                    Dim addressID = If(foreignKeyBox.ForeignKey Is Entity.FK_ShipToAddress, Entity.ShipToAddressID, Entity.BillToAddressID)
                    _addressLookupPopup.Show(foreignKeyBox, CurrentRow.GetValue(addressID), customerID.Value)
                End If
            End If
        End Sub

        Public ReadOnly Property SalesOrderId() As Integer
            Get
                Return Entity.SalesOrderID(0).Value
            End Get
        End Property

        Public ReadOnly Property IsNew() As Boolean
            Get
                Return SalesOrderId < 1
            End Get
        End Property

        Public Overrides Function SubmitInput(Optional focusToErrorInput As Boolean = True) As Boolean
            If Not MyBase.SubmitInput(focusToErrorInput) Then
                Return False
            End If

            Dim detailsPresenter = _subFormBinding(CurrentRow).DataPresenter
            Return detailsPresenter.SubmitInput(focusToErrorInput)
        End Function

        Public Async Function SaveToDb(ct As CancellationToken) As Task(Of System.Nullable(Of Integer))
            If IsNew Then
                Return Await App.ExecuteAsync(Function(db) db.CreateSalesOrderAsync(DataSet, ct))
            Else
                Await App.ExecuteAsync(Function(db) db.UpdateSalesOrderAsync(DataSet, ct))
                Return Nothing
            End If
        End Function
    End Class
End Class
